// supabase_service.proto - Supabase 数据库与向量服务 gRPC 接口
// 提供统一的 Supabase 访问能力，包括数据库操作和 pgvector 向量搜索
syntax = "proto3";

package isa.supabase;

option go_package = "github.com/isa-cloud/isa_cloud/api/proto/supabase";

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// ========================================
// Supabase 数据库与向量服务
// ========================================
// Supabase 提供：
// - PostgreSQL 数据库 (完整 SQL 支持)
// - pgvector 向量搜索 (语义搜索、RAG)
// - PostgREST API (自动生成 REST 接口)
// - 实时订阅 (Database changes)
// - 用户认证 (GoTrue Auth)
// 
// 本服务封装 Supabase 的核心功能，提供 gRPC 接口

service SupabaseService {
    // ========== 数据库操作 (Database CRUD) ==========
    
    // Query - 查询数据 (SELECT)
    rpc Query(QueryRequest) returns (QueryResponse);
    
    // Insert - 插入数据
    rpc Insert(InsertRequest) returns (InsertResponse);
    
    // Update - 更新数据
    rpc Update(UpdateRequest) returns (UpdateResponse);
    
    // Delete - 删除数据
    rpc Delete(DeleteRequest) returns (DeleteResponse);
    
    // Upsert - 插入或更新 (INSERT ON CONFLICT UPDATE)
    rpc Upsert(UpsertRequest) returns (UpsertResponse);
    
    // RPC - 调用 PostgreSQL 函数/存储过程
    rpc ExecuteRPC(RPCRequest) returns (RPCResponse);
    
    // ========== 向量操作 (Vector Operations - pgvector) ==========
    
    // UpsertEmbedding - 插入或更新向量嵌入
    rpc UpsertEmbedding(UpsertEmbeddingRequest) returns (UpsertEmbeddingResponse);
    
    // SimilaritySearch - 向量相似度搜索
    rpc SimilaritySearch(SimilaritySearchRequest) returns (SimilaritySearchResponse);
    
    // HybridSearch - 混合搜索 (全文 + 向量)
    rpc HybridSearch(HybridSearchRequest) returns (HybridSearchResponse);
    
    // DeleteEmbedding - 删除向量
    rpc DeleteEmbedding(DeleteEmbeddingRequest) returns (DeleteEmbeddingResponse);
    
    // ========== 批量操作 (Batch Operations) ==========
    
    // BatchInsert - 批量插入
    rpc BatchInsert(BatchInsertRequest) returns (BatchInsertResponse);
    
    // BatchUpsertEmbeddings - 批量插入向量
    rpc BatchUpsertEmbeddings(BatchUpsertEmbeddingsRequest) returns (BatchUpsertEmbeddingsResponse);
    
    // ========== 健康检查 ==========
    
    // HealthCheck - 服务健康检查
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ========================================
// 数据库操作消息定义
// ========================================

// QueryRequest - 查询请求
message QueryRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;              // 表名
    string select = 3;             // 选择字段 (e.g., "*", "id,name,email")
    string filter = 4;             // 过滤条件 (PostgREST 语法: "age.gte.18,status.eq.active")
    string order = 5;              // 排序 (e.g., "created_at.desc")
    int32 limit = 6;               // 限制返回数量
    int32 offset = 7;              // 偏移量
    bool count = 8;                // 是否返回总数
}

// QueryResponse - 查询响应
message QueryResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated google.protobuf.Struct data = 2;  // 查询结果 (JSON 数组)
    int32 count = 3;                           // 总数 (如果 count=true)
}

// InsertRequest - 插入请求
message InsertRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    repeated google.protobuf.Struct data = 3;  // 要插入的数据 (JSON 数组)
    bool return_data = 4;                      // 是否返回插入的数据
}

// InsertResponse - 插入响应
message InsertResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated google.protobuf.Struct data = 2;  // 插入的数据 (如果 return_data=true)
    int32 count = 3;                           // 插入的行数
}

// UpdateRequest - 更新请求
message UpdateRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    google.protobuf.Struct data = 3;           // 要更新的数据 (JSON)
    string filter = 4;                         // 过滤条件 (PostgREST 语法)
    bool return_data = 5;                      // 是否返回更新后的数据
}

// UpdateResponse - 更新响应
message UpdateResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated google.protobuf.Struct data = 2;  // 更新后的数据
    int32 count = 3;                           // 更新的行数
}

// DeleteRequest - 删除请求
message DeleteRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    string filter = 3;                         // 过滤条件 (PostgREST 语法)
    bool return_data = 4;                      // 是否返回删除的数据
}

// DeleteResponse - 删除响应
message DeleteResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated google.protobuf.Struct data = 2;  // 删除的数据
    int32 count = 3;                           // 删除的行数
}

// UpsertRequest - 插入或更新请求
message UpsertRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    repeated google.protobuf.Struct data = 3;  // 数据 (JSON 数组)
    string on_conflict = 4;                    // 冲突字段 (e.g., "id" or "email")
    bool return_data = 5;                      // 是否返回数据
}

// UpsertResponse - 插入或更新响应
message UpsertResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated google.protobuf.Struct data = 2;  // 数据
    int32 count = 3;                           // 受影响的行数
}

// RPCRequest - RPC 调用请求 (PostgreSQL 函数)
message RPCRequest {
    common.RequestMetadata metadata = 1;
    
    string function_name = 2;                  // 函数名
    google.protobuf.Struct params = 3;         // 参数 (JSON)
}

// RPCResponse - RPC 调用响应
message RPCResponse {
    common.ResponseMetadata metadata = 1;
    
    google.protobuf.Value data = 2;            // 返回值 (可以是任意 JSON)
}

// ========================================
// 向量操作消息定义 (pgvector)
// ========================================

// UpsertEmbeddingRequest - 插入/更新向量请求
message UpsertEmbeddingRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 向量表名
    string id = 3;                             // 文档 ID
    repeated float embedding = 4;              // 向量 (e.g., 1536 维的 OpenAI embedding)
    google.protobuf.Struct metadata_json = 5;  // 元数据 (JSON: title, content, tags, etc.)
}

// UpsertEmbeddingResponse - 插入/更新向量响应
message UpsertEmbeddingResponse {
    common.ResponseMetadata metadata = 1;
    
    string id = 2;                             // 文档 ID
    bool success = 3;                          // 是否成功
}

// SimilaritySearchRequest - 相似度搜索请求
message SimilaritySearchRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 向量表名
    repeated float query_embedding = 3;        // 查询向量
    int32 limit = 4;                           // 返回数量 (默认 10)
    string filter = 5;                         // 元数据过滤 (PostgREST 语法, e.g., "category.eq.tech")
    string metric = 6;                         // 距离度量 (cosine, l2, inner_product)
    float threshold = 7;                       // 相似度阈值 (0-1)
}

// SimilaritySearchResponse - 相似度搜索响应
message SimilaritySearchResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated VectorSearchResult results = 2;   // 搜索结果
}

// VectorSearchResult - 向量搜索结果
message VectorSearchResult {
    string id = 1;                             // 文档 ID
    float similarity = 2;                      // 相似度分数 (0-1)
    google.protobuf.Struct metadata = 3;       // 元数据
}

// HybridSearchRequest - 混合搜索请求 (全文 + 向量)
message HybridSearchRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    string text_query = 3;                     // 文本查询 (全文搜索)
    repeated float vector_query = 4;           // 向量查询
    int32 limit = 5;                           // 返回数量
    float text_weight = 6;                     // 文本权重 (0-1, 默认 0.5)
    float vector_weight = 7;                   // 向量权重 (0-1, 默认 0.5)
    string filter = 8;                         // 元数据过滤
}

// HybridSearchResponse - 混合搜索响应
message HybridSearchResponse {
    common.ResponseMetadata metadata = 1;
    
    repeated VectorSearchResult results = 2;   // 搜索结果
}

// DeleteEmbeddingRequest - 删除向量请求
message DeleteEmbeddingRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    string id = 3;                             // 文档 ID
}

// DeleteEmbeddingResponse - 删除向量响应
message DeleteEmbeddingResponse {
    common.ResponseMetadata metadata = 1;
    
    bool success = 2;                          // 是否成功
}

// ========================================
// 批量操作消息定义
// ========================================

// BatchInsertRequest - 批量插入请求
message BatchInsertRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    repeated google.protobuf.Struct data = 3;  // 数据数组
    int32 batch_size = 4;                      // 每批大小 (默认 100)
}

// BatchInsertResponse - 批量插入响应
message BatchInsertResponse {
    common.ResponseMetadata metadata = 1;
    
    int32 total_count = 2;                     // 总插入数
    int32 success_count = 3;                   // 成功数
    int32 error_count = 4;                     // 失败数
    repeated string errors = 5;                // 错误消息
}

// BatchUpsertEmbeddingsRequest - 批量插入向量请求
message BatchUpsertEmbeddingsRequest {
    common.RequestMetadata metadata = 1;
    
    string table = 2;                          // 表名
    repeated EmbeddingData embeddings = 3;     // 向量数组
}

// EmbeddingData - 向量数据
message EmbeddingData {
    string id = 1;                             // 文档 ID
    repeated float embedding = 2;              // 向量
    google.protobuf.Struct metadata = 3;       // 元数据
}

// BatchUpsertEmbeddingsResponse - 批量插入向量响应
message BatchUpsertEmbeddingsResponse {
    common.ResponseMetadata metadata = 1;
    
    int32 total_count = 2;                     // 总数
    int32 success_count = 3;                   // 成功数
    int32 error_count = 4;                     // 失败数
    repeated string errors = 5;                // 错误消息
}

// ========================================
// 健康检查
// ========================================

// HealthCheckRequest - 健康检查请求
message HealthCheckRequest {
    common.RequestMetadata metadata = 1;
}

// HealthCheckResponse - 健康检查响应
message HealthCheckResponse {
    common.ResponseMetadata metadata = 1;
    
    bool healthy = 2;                          // 是否健康
    string supabase_status = 3;                // Supabase 连接状态
    string postgres_version = 4;               // PostgreSQL 版本
    bool pgvector_enabled = 5;                 // pgvector 是否启用
}

